owrx/source/perseussdr.py

```diff
@@ -0,0 +1,31 @@
+from .direct import DirectSource
+from owrx.command import Flag, Option
+
+
+#
+# perseustest -s 768000 -u 0 -f 14150000 -r-|csdr convert_s8_f|nmux --bufsize 192512 --bufcnt 260 --port 35989 --address 127.0.0.1
+# perseustest -a -t0    -o -
+# perseustest -d 9 -a -t 100000 -o -  -s 768000 -u 0 -f 14150000 -r-
+
+class PerseussdrSource(DirectSource):
+    def getCommandMapper(self):
+        return super().getCommandMapper().setBase("perseustest -p -d -1 -a -t 100000 -o -  ").setMappings(
+            {
+                "samp_rate": Option("-s"),
+                "tuner_freq": Option("-f"),
+                "rf_gain": Option("-u"),
+                "lna_gain": Option("-g"),
+                "rf_amp": Option("-x"),
```

Comment:
I don't see `-g` or `-x` in https://github.com/Microtelecom/libperseus-sdr/blob/v0.7.5/perseustest.c, can you confirm they exist?

Reply:
This is is an open point.
Perseus has several controls:

attenuator
analog band filters
ADC dithering
ADC preamp

Initially, I thought that control could be implemented in the GUI.
There is such way?

Reply:
I don't think such controls should be in the receiver itself, since they can be a point of conflict. They should be set as best as possible in the profiles, and there may be a separate interface in the future to help with that.

Still, that won't work if those parameters aren't exposed through `perseustest`.

Reply:
also, i just noticed that the `-u` flag was added after version 0.7.5 as well. Was there a specific reason to go with 0.7.5?

Reply:
Because is quite an old stable release.
I understand that the only way (at the moment) to set that auxiliary parameters is using the static config, so I'll put out a new release of libperseus-sdr, so we can take into account all parameters.

Reply:
well, to me it seems like the `perseustest` utility isn't really designed for that purpose, so it seems to be lacking in some ways. I think it will work, but it'll be limited.

To fully unlock all the bells and whistles, I guess there needs to be some other piece of software that fits the purpose. Somewhat like the rtl_connector, but it may even be more reasonable to get started on a soapy driver (I couldn't find one right now) and use that with the soapy_connector. That would allow the Perseus to be used in a lot more applications, too.

Reply:
perseustest is maintained by me, as the whole libperseus-sdr, so it is just a matter to add that missing options (BTW I already did that).

On the other side all other SDR software that supports Perseus on Linux (gqrx, SDRAngel, gr-osmosdr, GNuRadio, Winrad and maybe I forgot something), obviously don't use perseustest or SoapySDR, BUT the native library (libperseus-sdr).

So, perseustest fits well the openwebrx model.

Reply:
But still is not clear to me why, in the hackrf class, there is
```
    def getEventNames(self):
        return super().getEventNames() + [
            "lna_gain",
            "rf_amp",
        ]
```
In the end this the perseus class:

```
from .direct import DirectSource
from owrx.command import Flag, Option


#
# In order to interface Perseus hardware, we resolve to use the
# perseustest utility that comes with libperseus-sdr support package.
# Below teh optios used are shown:
#
# -p output    I/Q samples as 32 bits floating point
# -d -1        suppress debug messages
# -a           don't test attenuators on startup
# -t 0         runs indefinitely
# -o -         output samples on stdout
#

class PerseussdrSource(DirectSource):
    def getCommandMapper(self):
        return super().getCommandMapper().setBase("perseustest -p -d -1 -a -t 0 -o -  ").setMappings(
            {
                "samp_rate": Option("-s"),
                "tuner_freq": Option("-f"),
                "rf_gain": Option("-u"),
                "adc_preamp": Option("-m"),
                "adc_dither": Option("-x"),
                "wideband": Option("-w"),
            }
        )

    def getEventNames(self):
        return super().getEventNames() + [
            "rf_gain",
            "adc_preamp",
            "adc_dither",
            "wideband",
        ]

    def getFormatConversion(self):
        # we are already returning I/Q samples as paire of 32 bits
        # floating points (option -p),so no need for further conversions
        return []

```

Reply:
I wouldn't mind interfacing with the library, but SoapySDR is far more convenient since it standardizes a lot of the stuff, hence the call. The OpenWebRX way has changed a bit ever since the introduction of the connectors. They allow direct interaction and allow switching profiles without the need of restarting any command-line commands. They're no must, that's why the DirectSource class still exists.

The `getEventNames()` should simply return the names of config / profile keys the class will respond to. There's a list of default ones in the `super().getEventNames()` call already, and if the class adds any more, it will need to add them here. The hackrf has mappings for the `lna_gain` and `rf_amp` keys, so it sets those up.

The class will receive calls to `onPropertyChange()` whenever a key is changed, which is already implemented in the DirectSource. It will simply be a `start()` / `stop()`, but you could do more refined things with the name and value if necessary.

I think you can leave out the `rf_gain` key, that is set up by default. `getFormatConversion()` in that form isn't necessary either, the parent class implements it in the same way already.